This Notebook will be used to create the 
# Capstone Project "Battle of the Neighborhoods". 
_It is part of the IBM Data Science Professional Certificate given through Coursera._

## 1. _Week 1_ Submission

In [1]:
import pandas as pd
import numpy as np

In [2]:
print("Hello Capstone Project Course!")

Hello Capstone Project Course!


## 2. _Week 3_ Submission

### 2.1 Scrapping

#### 2.1.1 Scrapping using Beautifoulsoup

Let's start by installing Beautifulsoup using conda install. 
As the package was not included in the standard conda environment I add the conda-forge channel to install Beautifulsoup 4.

In [3]:
conda config --add channels conda-forge


Note: you may need to restart the kernel to use updated packages.


In [4]:
conda install beautifulsoup4 bs4

Solving environment: ...working... done

# All requested packages already installed.


Note: you may need to restart the kernel to use updated packages.


In [5]:
conda update -n base -c defaults conda

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\arnau\anaconda3

  added / updated specs:
    - conda


The following packages will be REMOVED:

  python_abi-3.8-1_cp38

The following packages will be SUPERSEDED by a higher-priority channel:

  conda              conda-forge::conda-4.9.2-py38haa244fe~ --> pkgs/main::conda-4.9.2-py38haa95532_0


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done

Note: you may need to restart the kernel to use updated packages.


Once the package was installed let's import it. 

In [6]:
from bs4 import BeautifulSoup

Furthermore, we will need to install and import *requests* to scrape and receive the Wikipedia page as an html file.

In [7]:
conda install requests

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\arnau\anaconda3

  added / updated specs:
    - requests


The following NEW packages will be INSTALLED:

  python_abi         conda-forge/win-64::python_abi-3.8-1_cp38

The following packages will be SUPERSEDED by a higher-priority channel:

  conda               pkgs/main::conda-4.9.2-py38haa95532_0 --> conda-forge::conda-4.9.2-py38haa244fe_0


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done

Note: you may need to restart the kernel to use updated packages.


In [8]:
import requests

Once we installed and imported all required packages we can start **scrapping** the specified Wikipedia page. 

In [9]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
source = requests.get(url).text
    
soup = BeautifulSoup(source, 'lxml')

soup

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"22ea192a-9866-424a-ae12-2215b8fdea01","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":979555370,"wgRevisionId":979555370,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communications in Ontario","P

As we can see the html file looks very messy and contains far more information then what we need for this assignment. Looking through the HTML file and having a look at the source code directly on the Wikipedia Page itself we can find out that the required information is contained in the *table* attribute within the HTML file.  

We'll use the **find_all** function from Beautifulsoup to extract only the elements contained within the *table* attribute.

In [10]:
tables = soup.find_all('table')
tables

[<table class="wikitable sortable">
 <tbody><tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighbourhood
 </th></tr>
 <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>
 <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>
 <tr>
 <td>M5A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Regent Park, Harbourfront
 </td></tr>
 <tr>
 <td>M6A
 </td>
 <td>North York
 </td>
 <td>Lawrence Manor, Lawrence Heights
 </td></tr>
 <tr>
 <td>M7A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Queen's Park, Ontario Provincial Government
 </td></tr>
 <tr>
 <td>M8A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M9A
 </td>
 <td>Etobicoke
 </td>
 <td>Islington Avenue, Humber Valley Village
 </td></tr>
 <tr>
 <td>M1B
 </td>
 <td>Scarborough
 </td>
 <td>Malvern, Rouge
 </td></tr>
 <tr>
 <td>M2B

This is already far better!  

But still the format is still not quite the one we would like to get. We can see that the Headers we will use for the Table are specified as *th* whereas the rows we are interested in are preceded by *td*.  

Let's start by extracting the Headers from the HTML file and save them inside a list called **cols**. To do so, we will first search for all elements that start with *th* and then loop through these elements, eliminate the HTML code ('<th>' and '</th>), and then append these elements inside the list called **cols**.

In [11]:
headers = tables[0].find_all('th')
cols = []

for header in headers: 
    cols.append(header.string.strip())
    
cols

['Postal Code', 'Borough', 'Neighbourhood']

The second step is now to do the same with the rows. We'll start by finding all elements starting with *td* then for each of these elements we strip them from the HTML artefacts and add them inside a list called **Canada_table**. 

In [12]:
Canada_data = tables[0].find_all('td')
Canada_table = []

for data in Canada_data:
    Canada_table.append(data.string.strip())

However, we want to have a list where each element represents the full row. We know that the table will need to contain 3 attributes: Postal Code, Borough, Neighbourhood. Therefore, each row will contain exactly 3 elements.  

To slice the list into a list of lists containing each 3 elements we'll create the list called **canada_sublists** where each 3 elements from **Canada_table** will be read as a separate list and added to the **canada_sublists**. 

In [13]:
canada_sublists = [Canada_table[i:i+3] for i in range(0, len(Canada_table), 3)]

canada_sublists

[['M1A', 'Not assigned', 'Not assigned'],
 ['M2A', 'Not assigned', 'Not assigned'],
 ['M3A', 'North York', 'Parkwoods'],
 ['M4A', 'North York', 'Victoria Village'],
 ['M5A', 'Downtown Toronto', 'Regent Park, Harbourfront'],
 ['M6A', 'North York', 'Lawrence Manor, Lawrence Heights'],
 ['M7A', 'Downtown Toronto', "Queen's Park, Ontario Provincial Government"],
 ['M8A', 'Not assigned', 'Not assigned'],
 ['M9A', 'Etobicoke', 'Islington Avenue, Humber Valley Village'],
 ['M1B', 'Scarborough', 'Malvern, Rouge'],
 ['M2B', 'Not assigned', 'Not assigned'],
 ['M3B', 'North York', 'Don Mills'],
 ['M4B', 'East York', 'Parkview Hill, Woodbine Gardens'],
 ['M5B', 'Downtown Toronto', 'Garden District, Ryerson'],
 ['M6B', 'North York', 'Glencairn'],
 ['M7B', 'Not assigned', 'Not assigned'],
 ['M8B', 'Not assigned', 'Not assigned'],
 ['M9B',
  'Etobicoke',
  'West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale'],
 ['M1C', 'Scarborough', 'Rouge Hill, Port Union, Highland Creek'],
 ['M

Once we have the two separate lists, **cols** containing the Column Headers of our DataFrame and the **canada_sublists** containing our rows, we can merge them into a DataFrame, that we will call **canada_df**.

In [14]:
canada_df = pd.DataFrame(canada_sublists, columns=cols)

canada_df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

As mentioned in the assignment we will have to get rid of the lines in the DataFrame that have Boroughs that are *Not assigned*.  
To do so we will use the combination of the **str.contains** and the **~** to search for the rows that contain "Not assigned" and then substract them from the rest of the DataFrame.

In [15]:
canada_df = canada_df[~canada_df['Borough'].str.contains("Not assigned")]

canada_df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Afterwards we still need to check if there are some Neighbourhoods that contain "Not assigned". 

In [16]:
canada_df_unasgn = canada_df [canada_df['Neighbourhood'].str.contains("Not assigned")]

canada_df_unasgn

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

As you can see not Neighbourhood contained "Not assigned" as the DataFrame is empty. Therefore, we won't need to replace the name of unassigned Neighbourhoods.

To conclude let us print the shape of the DataFrame we could extract. 

In [17]:
print(canada_df.shape)

(103, 3)


#### 2.1.2 Scraping with _Pandas_

To scrape the Wikipedia page we can also use the *read_html* method integrated in Pandas. This method returns HTML tables into a list of DataFrame objects.

In [18]:
df_list = pd.read_html(url)

Let's check how many HTML tables were found on the Wikipedia Page.

In [19]:
len(df_list)

3

We can see that the Dataframe contains 3 objects. However, the one we are interested in is the first one, the first table that appears on the Wikipedia page.  

Therefore, we will read the first DataFrame object into a new DataFrame named **canada_df_pandas**.

In [20]:
canada_df_pandas = df_list[0]

canada_df_pandas

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

As you can see we already have our table with the correct headers and rows without having to clean the data. However, we still need to delete the rows that have a "Not assigned" Borough.  

Let's proceed like with during the BeautifulSoup chapter. We'll use the ***str.contains*** method combined with the ***~*** to get rid off each row that as an unassigned Borough. 


In [21]:
canada_df_pandas = canada_df_pandas[~canada_df_pandas['Borough'].str.contains("Not assigned")]

canada_df_pandas

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Now let's check if there are Neighbourhoods that are unassigned. Like with the unassigned Boroughs we can use the ***str.contains*** method. But this time we won't erase the rows that contain "Not assigned", we want only to see in which instances Neighbourhoods are not assigned. 

In [22]:
canada_df_pandas_unasgn = canada_df_pandas[canada_df_pandas['Neighbourhood'].str.contains("Not assigned")]

canada_df_pandas_unasgn

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

As you can see, no Neighbourhood is unassigned.  

Therefore, let's end the exercise by printing out the shape of the obtained DataFrame **canada_df_pandas**.

In [23]:
print(canada_df_pandas.shape)

(103, 3)


### 2.2 Latitude and Longitude Data

To determine the Latitude and Longitude of the different parts of Toronto we will use the CSV file provided in the assignement session as the geocoder package refused to work.  

Let's start by reading the CSV into a Pandas DataFrame.

In [24]:
csv_url = "https://cocl.us/Geospatial_data"

lat_long = pd.read_csv(csv_url)

lat_long.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [25]:
canada_df = canada_df.merge(lat_long, on=['Postal Code'])  


In [26]:
canada_df.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

### 2.3 Exploration and Clustering

Before starting this section we will install and import the different packages need for the following sections.

In [27]:
!pip install folium
import folium

!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import numpy as np # library to handle data in a vectorized manner

print('packages installed and imported')

import json

packages installed and imported


Let's start by getting the Latitude and Longitude of Toronto, this will help us create the map of Toronto with Folium.

In [28]:
address = 'Toronto'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are: {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are: 43.6534817, -79.3839347.


Now let's use the Latitude and Longitude data we saved in our canada_df DataFrame to mark the Toronto map with the different Neighborhoods in Toronto.

In [29]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(canada_df['Latitude'], canada_df['Longitude'], canada_df['Borough'], canada_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

As mentioned in the Assignment we will restrict our analyses to the Boroughs that only include the word 'Toronto'. In that effect we will create a new DataFrame called *toronto_data*.  

Let's print the output of this new DataFrame.

In [30]:
toronto_data = canada_df[canada_df['Borough'].str.contains("Toronto")].reset_index(drop=True)

toronto_data.rename(columns = {'Neighbourhood':'Neighborhood'}, inplace = True)

toronto_data

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                 North Toronto West,  Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

As most of the Neighborhoods are located in the Downtown Toronto Borough we will center the map on this Borough. To get the Latitude and Longitude of Downtown Toronto we will use the Nominatim geolocator. 

In [31]:
address_old_toronto = 'Downtown Toronto'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address_old_toronto)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are: {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are: 43.6563221, -79.3809161.


As we have now the Latitude and Longitude of Downtown Toronto let's create the 'restricted' map with only the Boroughs having 'Toronto' in their name. 

In [32]:
map_downtown_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown_toronto)  
    
map_downtown_toronto

**Foursquare API**  

Now we will use the Foursquare API to get more information about the venues in the identified Neighborhoods. Furthermore, we will use the information about the Category of venues provided by the Foursquare API.  

For this we will create a function called getNearbyVenues. 

In [33]:
CLIENT_ID = 'NIDDKXAR0Y1JRGEEWFFQHC234U455BN21PQDFE3EFDRP1GJC' # your Foursquare ID
CLIENT_SECRET = '2FO3JXOSKZJLNJKOYE3L120JFVE1T2S5UKDCJC5NGP2PKVST' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: NIDDKXAR0Y1JRGEEWFFQHC234U455BN21PQDFE3EFDRP1GJC
CLIENT_SECRET:2FO3JXOSKZJLNJKOYE3L120JFVE1T2S5UKDCJC5NGP2PKVST


In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [35]:
toronto_venues= getNearbyVenues(names=toronto_data['Neighborhood'], latitudes=toronto_data['Latitude'], longitudes=toronto_data['Longitude'])

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West,  Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport


Once we looped through all the identified Neighborhoods, we can start by analyzing the venue categories that are present in the different Neighborhoods and put them into a new DataFrame called *toronto_venues*.  

Let's see how many venues were identified.

In [36]:
print(toronto_venues.shape)

toronto_venues.head()

(1639, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

We can see that we identified 1639 venues accross the Boroughs that contain 'Toronto'.  

To understand how many venues were indentified by Neighborhood let's create a *toronto_venues_grouped* DataFrame by grouping the *toronto_venues* DataFrame. 

In [37]:
toronto_venues_grouped = pd.DataFrame(toronto_venues.groupby('Neighborhood')['Venue Category'].count())

toronto_venues_grouped.head()

Venue Category
Neighborhood                                                      
Berczy Park                                                     55
Brockton, Parkdale Village, Exhibition Place                    25
Business reply mail Processing Centre, South Ce...              16
CN Tower, King and Spadina, Railway Lands, Harb...              17
Central Bay Street                                              68

Let's print how many unique categories of venues are present in our 'filtered' Toronto Boroughs.

In [38]:
print('There are {} uniques categories in Toronto.'.format(len(toronto_venues['Venue Category'].unique())))

There are 240 uniques categories in Toronto.


Now that we have some information about the categories of venues present in the different Neighborhoods we can start understanding what kind of Neighborhoods we are dealing with.

For this we will try to map the frequency of these 240 unique venue categories and group them by Neighborhood.

In [78]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head(10)

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   
5            0                  0        0                   0             0   
6            0                  0        0                   0             0   
7            0                  0        0                   0             0   
8            0                  0        0                   0             0   
9            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   
5               0                0                 0                    0   
6               0                0                 0                    0   
7               0                0                 0                    0   
8               0                0                 0                    0   
9               0                0                 0                    0   

   Antique Shop  ...  Theater  Theme Restaurant  Toy / Game Store  Trail  \
0             0  ...        0                 0                 0      0   
1             0  ...        0                 0                 0      0   
2             0  ...        0                 0                 0      0   
3             0  ...        0                 0                 0      0   
4             0  ...        0                 0                 0      0   
5             0  ...        0                 0                 0      0   
6             0  ...        0                 0                 0      0   
7             0  ...        0                 0                 0      0   
8             0  ...        0                 0                 0      0   
9             0  ...        0                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   
5              0                              0                 0   
6              0                              0                 0   
7              0                              0                 0   
8              0                              0                 0   
9              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  
0                      0         0              0  
1                      0         0              0  
2                      0         0              0  
3                      0         0              0  
4                      0         0              0  
5                      0         0              0  
6                      0         0              0  
7                      0         0              0  
8                      0         0              0  
9                      0         0              0  

[10 rows x 240 columns]

In [40]:
toronto_onehot.shape

(1639, 240)

In [41]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()

toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.014706   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.026667   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.021739   
20                        Moore Park, Summerhill East     0.000000   
21                 North Toronto West,  Lawrence Park     0.055556   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.030303   
24                          Regent Park, Harbourfront     0.022222   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.028571   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010417   
32                                    Studio District     0.027027   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.022727   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.029412   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.058824            0.058824      0.058824   
4            0.000000  0.000000            0.000000      0.000000   
5            0.000000  0.000000            0.000000      0.000000   
6            0.013333  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.000000  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.000000  0.000000            0.000000      0.000000  

In [42]:
toronto_grouped.shape

(39, 240)

As we see we have a row by Neighborhood and a frequency value for each venue category.  

Now that we have this DataFrame we can group the top 5 venue categories by Neighborhood.

In [43]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.04
2      Farmers Market  0.04
3              Bakery  0.04
4  Seafood Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1       Nightclub  0.08
2     Coffee Shop  0.08
3  Breakfast Spot  0.08
4   Grocery Store  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
              venue  freq
0       Pizza Place  0.06
1     Auto Workshop  0.06
2        Comic Shop  0.06
3  Recording Studio  0.06
4        Restaurant  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0       Airport Lounge  0.12
1      Airport Service  0.12
2  Rental Car Location  0.06
3                  Bar  0.06
4     Sculpture Garden  0.06


----Central Bay Street----
                venue  freq
0         Coff

In order to get a DataFrame that list the 10 top most common venues by Neighborhood we'll start by creating a function called 'return_most_common_venues'.  

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [45]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...                  Park   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Farmers Market                Bakery              Beer Bar   
1           Coffee Shop        Breakfast Spot             Nightclub   
2      Recording Studio            Restaurant    Light Rail Station   
3       Airport Service                 Plane       Harbor / Marina   
4                  Café        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Cocktail Bar           Cheese Shop    Seafood Restaurant   
1          Climbing Gym                   Bar             Bookstore   
2         Auto Workshop  Fast Food Restaurant        Farmers Market   
3              Boutique         Boat or Ferry   Rental Car Location   
4      Department Store   Japanese Restaurant       Thai Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant         Grocery Store                    Pub  
1         Burrito Place            Restaurant             Playground  
2         Burrito Place           Pizza Place                Butcher  
3                   Bar         Historic Site            Coffee Shop  
4          Burger Joint       Bubble Tea Shop            Salad Place

### 2.4 Cluster the Neighbourhoods

Now that we have a DataFrame that gives us some information about each Neighborhood analyzed we can start by running the k-means clustering to seperate the Neighborhoods by clusters.  

We will set the number of clusters at 5.  

In [46]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

As we now have an array of clusters let's add this information to the neighborhood_venues_sorted DataFrame. Once we did this we can join this DataFram with the general *toronto_data*.  

In [47]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0           Coffee Shop   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               1     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery                   Pub   
1           Yoga Studio                  Café              Beer Bar   
2        Clothing Store                  Café       Bubble Tea Shop   
3                  Café          Cocktail Bar            Restaurant   
4                   Pub                 Trail              Dive Bar   

  5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
0                  Café               Theater          Breakfast Spot   
1         Smoothie Shop    Italian Restaurant          Sandwich Place   
2        Cosmetics Shop   Japanese Restaurant  Furniture / Home Store   
3             Gastropub              Beer Bar     American Restaurant   
4          Dessert Shop    Dim Sum Restaurant                   Diner   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Shoe Store   Distribution Center      Electronics Store  
1   Distribution Center            Restaurant                  Diner  
2                 Hotel           Pizza Place              Bookstore  
3   Japanese Restaurant    Seafood Restaurant         Clothing Store  
4        Discount Store   Distribution Center          Women's Store

Now let's visualize the Clusters on the map. 

In [48]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### 2.5 Examining the Clusters

Now that we have 5 Clusters, we can dive in and analyze what characterizes these Clusters. Let's slice the created DataFrame by Clusters. 

**First Cluster**

In [49]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0         Grocery Store   
8   Downtown Toronto               0           Coffee Shop   
9       West Toronto               0              Pharmacy   
10  Downtown Toronto               0           Coffee Shop   
11      West Toronto               0                   Bar   
12      East Toronto               0      Greek Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14      West Toronto               0                  Café   
15      East Toronto               0                  Park   
16  Downtown Toronto               0           Coffee Shop   
17      East Toronto               0           Coffee Shop   
20   Central Toronto               0                  Park   
22      West Toronto               0                   Bar   
23   Central Toronto               0           Coffee Shop   
24   Central Toronto               0        Sandwich Place   
25      West Toronto               0             Gift Shop   
26   Central Toronto               0           Pizza Place   
27  Downtown Toronto               0                  Café   
28      West Toronto               0                  Café   
30  Downtown Toronto               0    Mexican Restaurant   
31   Central Toronto               0           Coffee Shop   
32  Downtown Toronto               0        Airport Lounge   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0           Coffee Shop   
36  Downtown Toronto               0           Coffee Shop   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0                  Park   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                   Park                         Bakery                   Pub   
1            Yoga Studio                           Café              Beer Bar   
2         Clothing Store                           Café       Bubble Tea Shop   
3                   Café                   Cocktail Bar            Restaurant   
5         Farmers Market                         Bakery              Beer Bar   
6                   Café                 Sandwich Place    Italian Restaurant   
7                   Café                           Park           Candy Store   
8                   Café                          Hotel            Restaurant   
9                 Bakery                  Grocery Store           Supermarket   
10              Aquarium                          Hotel                  Café   
11           Coffee Shop          Vietnamese Restaurant            Restaurant   
12           Coffee Shop             Italian Restaurant            Restaurant   
13                 Hotel                     Restaurant                  Café   
14           Coffee Shop                 Breakfast Spot             Nightclub   
15           Pizza Place                 Ice Cream Shop                   Pub   
16            Restaurant                          Hotel                  Café   
17                Bakery                      Gastropub               Brewery   
20                 Hotel                 Breakfast Spot               Dog Run   
22       Thai Restaurant                           Café    Mexican Restaurant   
23        Clothing Store                    Yoga Studio            Bagel Shop   
24                  Café                    Coffee Shop        History Museum   
25        Breakfast Spot                     Restaurant                   Bar   
26          Dessert Shop                 

From the above table we can determine that the first cluster groups the Neighborhoods that have mostly Coffe Shops, Restaurants, and Bars. Therefore, let's rename this Cluster "Eat and Drink". 

**Second Cluster**

In [50]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4       East Toronto               1     Health Food Store   
21   Central Toronto               1                  Park   
33  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4                    Pub                 Trail              Dive Bar   
21      Sushi Restaurant         Jewelry Store                 Trail   
33                 Trail          Tennis Court            Playground   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4           Dessert Shop    Dim Sum Restaurant                 Diner   
21    Mexican Restaurant   Dumpling Restaurant            Donut Shop   
33      Doner Restaurant               Dog Run            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4         Discount Store   Distribution Center          Women's Store  
21      Doner Restaurant               Dog Run               Dive Bar  
33              Dive Bar          Dance Studio    Distribution Center

This Neighbourhood has mainly Trail and Playground venues, as such we could see this Neighbourhood as being more "Outdoor" activity focused. An aptly name would be "Outdoor Fun". 

**Third Cluster**

In [51]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29  Central Toronto               2                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29            Playground         Women's Store         Deli / Bodega   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
29  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29      Doner Restaurant               Dog Run               Dive Bar

This part of Central Toronto has more Parks and Swim Schools, therefore we can deduced that it's a perfect Neighborhood for people interested in Swimming, let's rename this one "Nature and Swim".

**Fourth Cluster**

In [52]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
18           Swim School              Bus Line  Eastern European Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18   Dumpling Restaurant            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18               Dog Run              Dive Bar    Distribution Center

We can see that in the 4th Cluster the most common venues are a combination of Park, Restaurants, and Outdoor activities such as Trail. This combination could be renamed "Nature, Fit & Eat".

**Fifth Cluster**

In [53]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               4              Dive Bar   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19           Music Venue                Garden         Women's Store   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
19         Deli / Bodega  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19            Donut Shop      Doner Restaurant                Dog Run

The 5th and last Cluster is a Neihbourhood located in Central Toronto that has the particularity to focus more on Music. Therefore, we could rename this Cluster "Music Lovers". 

**Conclusion**  

As we analyzed all the Clusters we can now rename them in the Map we created to reflect better the type of Neighbourhoods we identified. 

In [54]:
new_cluster_names = ('Eat and Drink', 'Outdoor Fun', 'Nature & Swim', 'Nature, Fit & Eat', 'Music Lovers')

In [68]:
toronto_merged['Cluster Labels'].replace({0: new_cluster_names[0], 1: new_cluster_names[1], 2: new_cluster_names[2], 3: new_cluster_names[3], 4: new_cluster_names[4]}, inplace = True)

toronto_merged


Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         The Beaches  43.676357 -79.293031   
5                                         Berczy Park  43.644771 -79.373306   
6                                  Central Bay Street  43.657952 -79.387383   
7                                            Christie  43.669542 -79.422564   
8                            Richmond, Adelaide, King  43.650571 -79.384568   
9                        Dufferin, Dovercourt Village  43.669005 -79.442259   
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
11                           Little Portugal, Trinity  43.647927 -79.419750   
12                       The Danforth West, Riverdale  43.679557 -79.352188   
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
15                     India Bazaar, The Beaches West  43.668999 -79.315572   
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
17                                    Studio District  43.659526 -79.340923   
18                                      Lawrence Park  43.728020 -79.388790   
19                                           Roselawn  43.711695 -79.416936   
20                                   Davisville North  43.712751 -79.390197   
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
22                      High Park, The Junction South  43.661608 -79.464763   
23                 North Toronto West,  Lawrence Park  43.715383 -79.405678   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                             Parkdale, Roncesvalles  43.648960 -79.456325   
26                                         Davisville  43.704324 -79.388790   
27                     University of Toronto, Harbord  43.662696 -79.400049   
28                                 Runnymede, Swansea  43.651571 -79.484450   
29                        Moore Park, Summerhill East  43.689574 -79.383160   
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
31  Summerhill West, Rathnelly, So

In order to keep the different color mapping for the Cluster Labels we will create a dictionary of the renamed Cluster Labels that will have values from 0 to 4.

In [66]:
d = {ni: indi for indi, ni in enumerate(set(toronto_merged['Cluster Labels']))}
numbers = [d[ni] for ni in toronto_merged['Cluster Labels']]

In [76]:
new_map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster:  ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[d.get(cluster)-1],
        fill=True,
        fill_color=rainbow[d.get(cluster)-1],
        fill_opacity=0.7).add_to(new_map_clusters)
       
new_map_clusters

Finally we have a map of the Toronto Boroughs that reflect the different types of Neighborhoods based on the most frequent venues.